fix multithread zero copy crash

rafiw opened the pull request:
sockperf used a global variable to call free_packet API.
This casued double free in VMA buffers and caused VMA to crash.
This fix changes the globals to be thread locals.

Signed-off-by: Rafi Wiener <rafiw@mellanox.com>

Commit:
fix multithread zero copy crash

sockperf used a global variable to call free_packet API.
This casued double free in VMA buffers and caused VMA to crash.
This fix changes the globals to be thread locals.

Signed-off-by: Rafi Wiener <rafiw@mellanox.com>

rosenbaumalex commented on src/Client.cpp:
```diff
@@ -472,6 +472,15 @@ ::initBeforeLoop()
 	if (rc == SOCKPERF_ERR_NONE)
 	{
+#ifdef USING_VMA_EXTRA_API
+		if (g_vma_api){
```
formatting: add space "if (...) {"

rosenbaumalex commented on src/Server.cpp:
```diff
@@ -358,6 +358,15 @@ void server_handler(int _fd_min, int _fd_max, int _fd_num) {
 {
 	if (p_info) {
+#ifdef USING_VMA_EXTRA_API
+		if (g_vma_api){
```
formatting: add space '...) {'

rosenbaumalex commented on src/Client.cpp:
```diff
@@ -472,6 +472,15 @@ ::initBeforeLoop()
 	{
+#ifdef USING_VMA_EXTRA_API
+		if (g_vma_api){
+			g_pkt_buf = new unsigned char [Message::getMaxSize()];
```
preferably use the MALLOC() macro with memory debug options. Why change?

rosenbaumalex commented on src/Defs.h:
```diff
@@ -385,10 +385,10 @@ extern TicksTime g_cycleStartTime;
-extern struct vma_packets_t* g_pkts;
-extern unsigned int g_pkt_index;
-extern unsigned int g_pkt_offset;
+extern __thread unsigned char* g_pkt_buf;
```
once using the '__thread' directive it would be better to:
1.  use MACRO definition. I'm guessing this will translate into different API's for different OS's
2.  move all parameters into a single structure for mapping VMA_EXTAR variable. Bringing the thread specific data into context has cost, let's get hit once for all 4 members

rafiw replied:
currently no need to support windows since EXTRA_API is under #if ! defined (WIN32) && ! defined (__FreeBSD__)
https://github.com/Mellanox/sockperf/blob/sockperf_v2/src/Defs.h#L97
i will change them all to one struct

Commit:
fix_comments

Signed-off-by: Rafi Wiener <rafiw@mellanox.com>

Commit:
wrap thread local inside function to improve performance

software.intel.com/en-us/blogs/2011/05/02/the-hidden-performance-cost-of-accessing-thread-local-variables

Signed-off-by: Rafi Wiener <rafiw@mellanox.com>

rafiw commented:
@avnerbh what do you think?

avnerbh commented on src/Client.cpp:
```diff
@@ -673,6 +682,11 @@ ::doHandler()
 
 		cleanupAfterLoop();
 	}
+#ifdef USING_VMA_EXTRA_API
```
Since you placed the new memory allocation inside _initBeforeLoop_ (which is appropriate), I am expecting that the new memory cleanup will be done inside _cleanupAfterLoop_
This will match our attitude for all other initialization and allocations that avoid messing the high level logic of _doHandler_ with low level code.

rafiw replied:
i agree but this can cause a potential memory leak if the init function fails and cleanupAfterLoop is not called

avnerbh replied:
in Sockperf we don't care for memory leaks that happen just before exit and we prefer readability.  At any case, this is O(1) leak

avnerbh commented on src/Defs.h:
```diff
@@ -384,11 +384,18 @@ extern TicksTime g_cycleStartTime;
-extern unsigned int g_pkt_index;
-extern unsigned int g_pkt_offset;
+#ifdef USING_VMA_EXTRA_API
+typedef struct zero_copy_thread_data {
```
_thread_ should not be part of the new type.  the TLS thing should be determined upon usage and is not appropriate in the type name.
(Also consider removing the the term _zero_copy_ from the type name.  this type is essentially a data-buffer regardless of its current specific usage.  If you want to relate it to vma, then you can call it vma_data_buffer)

rafiw replied:
changed to vma_data_buffer

avnerbh commented on src/Server.cpp:
```diff
@@ -358,6 +358,15 @@ void server_handler(int _fd_min, int _fd_max, int _fd_num) {
+				return;
+			}
+		}
+#endif
```
this low level code makes the _server_handler_ function to be ugly since without it it looks clean - dealing only with high level logic that can be understood at a glance.
You should create new general purpose initialization function and place your code inside it

rafiw replied:
changed to functions in common.h

avnerbh commented on src/Server.cpp:
```diff
@@ -391,6 +400,11 @@ void server_handler(handler_info *p_info)
+		if (g_zero_data.pkt_buf) {
+			FREE(g_zero_data.pkt_buf);
+		}
+#endif
```
the equivalent comment --> should go to general purpose cleanup function

rafiw replied:
ok

avnerbh commented on src/os_abstract.h:
```diff
@@ -136,7 +137,8 @@ void* win_set_timer(void *p_timer);
 #include <endian.h>
-
+#define SP_FORCE_INLINE inline __attribute__ ((always_inline))
+#define THREAD_LOCAL	__thread
```
Did you verify that your code compiles in all supported platforms and not only gcc and windows (for example BSD)?

rafiw replied:
yes

avnerbh reviewed (changes requested):
Hi Rafi, I pointed several things that jumped into my eyes.  I assume that you did with Alex the low level review.

PS, please feel free to ping me directly in case I am not responding to you.  Sometimes inbox rules hide it from me

igor-ivanov commented:
@rafiw could you please rebase

rafiw commented:
#89  will be used instead
